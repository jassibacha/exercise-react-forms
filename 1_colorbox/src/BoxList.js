import React, { useState } from 'react';
import NewBoxForm from './NewBoxForm';
import Box from './Box';
import { v4 as uuid } from 'uuid';

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);

    const renderBoxes = () => {
        return (
            <div>
                {boxes.map((box) => (
                    <Box
                        width={box.width}
                        height={box.height}
                        backgroundColor={box.backgroundColor}
                        id={box.id}
                        key={box.id}
                        deleteBox={() => deleteBox(box.id)}
                    />
                ))}
            </div>
        );
    };

    const deleteBox = (boxId) => {
        // Run setBoxes again
        setBoxes((boxes) =>
            // Filter the boxes array from state
            boxes.filter((box) => {
                // If the ID matches, remove it!
                return box.id !== boxId;
            })
        );
    };

    /** Add new item object to cart. */
    const addBox = (box) => {
        let newBox = { ...box, id: uuid() };
        setBoxes((boxes) => [...boxes, newBox]);
    };

    return (
        <div className="Boxes">
            <NewBoxForm addBox={addBox} />
            {renderBoxes()}
        </div>
    );
};

export default BoxList;
