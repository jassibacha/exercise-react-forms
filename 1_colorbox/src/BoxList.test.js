import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';

it('renders without crashing', function () {
    render(<BoxList />);
});

it('matches snapshot', function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('can add a new box', function () {
    const { getByLabelText, queryByText, getByText, toHaveStyle } = render(
        <BoxList />
    );

    // check there's no Delete button
    expect(queryByText('X')).not.toBeInTheDocument();

    const widthInput = getByLabelText('Width:');
    const heightInput = getByLabelText('Height:');
    const bgColorInput = getByLabelText('Background Color:');
    const submitBtn = getByText('Add a new box!');

    // fill out the form
    fireEvent.change(widthInput, { target: { value: '100' } });
    fireEvent.change(heightInput, { target: { value: '150' } });
    fireEvent.change(bgColorInput, { target: { value: '#FF0000' } });
    fireEvent.click(submitBtn);

    // box exists
    const deleteBtn = queryByText('X');
    expect(deleteBtn).toBeInTheDocument();
    // check if the box has the correct background color
    expect(deleteBtn.parentElement).toHaveStyle('background-color: #FF0000');
});

it('can delete a box', function () {
    const { getByLabelText, getByText, queryByText } = render(<BoxList />);

    const widthInput = getByLabelText('Width:');
    const heightInput = getByLabelText('Height:');
    const bgColorInput = getByLabelText('Background Color:');
    const submitBtn = getByText('Add a new box!');

    // fill out the form
    fireEvent.change(widthInput, { target: { value: '100' } });
    fireEvent.change(heightInput, { target: { value: '150' } });
    fireEvent.change(bgColorInput, { target: { value: '#FF0000' } });
    fireEvent.click(submitBtn);

    // box exists
    const deleteBtn = getByText('X');
    expect(deleteBtn).toBeInTheDocument();

    // delete the box
    fireEvent.click(deleteBtn);

    // box is gone!
    expect(queryByText('X')).not.toBeInTheDocument();
});
