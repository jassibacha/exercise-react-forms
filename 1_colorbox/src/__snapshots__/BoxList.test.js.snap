// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Boxes"
  >
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value=""
      />
      <label
        for="backgroundColor"
      >
        Background Color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="color"
        value="#000000"
      />
      <button>
        Add a new box!
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
